src/api/filter.rs

```diff
@@ -0,0 +1,208 @@
+}
+
+// tests the functionality of a filter by running it on alerts in database
+async fn test_filter(
+    client: web::Data<Client>, 
+    catalog: String, 
+    filter: Filter
+) -> Result<(), mongodb::error::Error> {
+    let collection: Collection<mongodb::bson::Document> = 
+        client.database(DB_NAME).collection(format!("{}_alerts", catalog).as_str());
+    
+    let result = collection.aggregate(filter.pipeline).await;
+    match result {
+        Ok(_) => {
+            return Ok(());
+        },
+        Err(e) => {
+            return Err(e);
+        }
+    }
+}
+
+// takes tested filter and builds the properly formatted bson document for the database
+fn build_filter_bson(
+    filter: Filter
+) -> Result<mongodb::bson::Document, mongodb::error::Error> {
+    // generate new object id
+    let id = mongodb::bson::oid::ObjectId::new();
+    let date_time = mongodb::bson::DateTime::now();
+    // TODO: how to manage multiple filters in each FILTER within the database
+    let database_filter_bson = doc! {
+        "_id": id,
+        "group_id": 41, // consistent with other test filters
+        "filter_id": filter.id,
+        "catalog": filter.catalog,
+        "permissions": filter.permissions,
+        "active": true,
+        "active_fid": filter.id,
+        "fv": [
+            // TODO: how to generate pipeline id's?
+            {
+                "fid": "some_pipeline_id",
+                "pipeline": filter.pipeline,
+                "created_at": date_time,
+            }
+        ],
+        "autosave": false,
+        "update_annotations": true,
+        "created_at": date_time,
+        "last_modified": date_time,
+    };
+    Ok(database_filter_bson)
+}
+
+// TODO: (filter id creation process)
+//      idea: the filter id creation process will be done by the programmer on the front end(?)
+//      idea: use authentication to help attribute the filter being added to the user adding it
+//            allows user to submit a filter to the database
+#[post("/filter")]
+pub async fn post_filter(
+    client: web::Data<Client>, 
+    body: web::Json<FilterSubmissionBody>
+) -> HttpResponse {
+    let body = body.clone();
+    // 1. grab user filter
+    let catalog = body.catalog.expect("catalog required");
+    let id = body.id.expect("filter id required");
+    let permissions = body.permissions.expect("filter permissions required");
+    let pipeline = body.pipeline.expect("filter pipeline required");
+    
+    // Test filter
+    
+    // 1. create production version of filter
+    let filter = build_test_filter(
+        catalog.clone(), id, permissions.clone(), pipeline.clone());
+    
+    // 2. perform test run to ensure no errors
+    match test_filter(client.clone(), catalog.clone(), filter).await {
+        Ok(()) => {},
+        Err(e) => {
+            println!("could not submit filter to database, got error: {}", e);
+            return HttpResponse::BadRequest().body(format!("Invalid filter submitted"));
```

Comment:
I wonder what kinds of errors we might get here. Would they all be due to a bad request?

Reply:
not necessarily no. We want to handle the type of error. But most will be. By design filters should not fail to be inserted in the DB except:
- there is an issue communicating with the DB, that is a proper failure and not a bad request
- the filter is incorrect, so we can't insert it. That's a bad request for sure